Support helm chart server separate liveness and readiness probe endpoints

alvinzhangdd opened the pull request:
### Change
Support helm chart server separate liveness and readiness probe endpoints

### Issue
https://github.com/apache/pinot/issues/11798

### Problem
Latest pinot helm chart doesn't support separate liveness and readiness probe endpoint for server ([link](https://github.com/apache/pinot/blob/master/helm/pinot/values.yaml#L311-L314)) (both use `server.probes.endpoint`)  but the service underneath supports it ([link](https://github.com/apache/pinot/pull/9031)) . 

### Proposal
Add supports for `server.probes.livenessEndpoint` and `server.probes.readinessEndpoint` to helm chart server sts template, but make it backward compatible by falling back to the existing `server.probes.endpoint`.

### Testing
Under `helm/pinot/`
* `helm dependency update`
* `helm install -f ./values.yaml pinot . -n pinot`. It can spin up the k8s pods locally.
* Edit values.yaml Test for separate liveness and readiness probe endpoint for server.
  * With livenessEndpoint: "/health/liveness", readinessEndpoint: "/health/readiness"
     * Liveness:   http-get http://:8097/health%3FcheckType=liveness delay=60s timeout=1s period=10s #success=1 #failure=3
     * Readiness:  http-get http://:8097/health%3FcheckType=readiness delay=60s timeout=1s period=10s #success=1 #failure=3
  * Without
     * Liveness: http-get http://:7443/health delay=120s timeout=1s period=20s #success=1 #failure=3
     * Readiness: http-get http://:7443/health delay=120s timeout=1s period=20s #success=1 #failure=3

walterddr reviewed (commented):
looks good here. but don't you need to also change the values.yaml to enable liveness endpoint? CC @xiangfu0

alvinzhangdd commented:
@walterddr 
I don't quite follow, do you mean I should change the values.yaml? Can you tell me where should I change?

I thought I should keep values.yaml unchanged, it is the person who want it enabled would set the following
```
server:
probes:
    endpoint: "/health"
    livenessEnabled: true
    livenessEndpoint: "/health/liveness"
    readinessEnabled: true
    readinessEndpoint: "/health/readiness"
```

Should I add those params livenessEndpoint, readinessEndpoint as optional params to README?

zhtaoxiang commented on helm/pinot/templates/server/statefulset.yaml:
```diff
@@ -88,15 +88,23 @@ spec:
           initialDelaySeconds: {{ .Values.probes.initialDelaySeconds }}
           periodSeconds: {{ .Values.probes.periodSeconds }}
           httpGet:
+            {{- if .Values.server.probes.livenessEndpoint }}
```
two suggestions:
1. change the property to `.Values.server.livenessProbe.endpoint`
2. change line 86 to {{- if or .Values.server.probes.livenessEnabled .Values.server.livenessProbe.endpoint }}

xiangfu0 replied:
+1, let's make sure we have the properties following the conventions.

alvinzhangdd replied:
Can you tell me why changing line 86 to if `Values.server.probes.livenessEnabled` OR `.Values.server.livenessProbe.endpoint` ?

Shouldn't if `Values.server.probes.livenessEnabled` be enough?

zhtaoxiang commented on helm/pinot/templates/server/statefulset.yaml:
```diff
@@ -88,15 +88,23 @@ spec:
           initialDelaySeconds: {{ .Values.probes.initialDelaySeconds }}
           periodSeconds: {{ .Values.probes.periodSeconds }}
           httpGet:
+            {{- if .Values.server.probes.readinessEndpoint }}
```
same here

walterddr commented:
> @walterddr I don't quite follow, do you mean I should change the values.yaml? Can you tell me where should I change?
> 
> I thought I should keep values.yaml unchanged, it is the person who want it enabled would set the following

good point. first of all the endpoints should be:
```
server:
probes:
    endpoint: "/health"
    livenessEnabled: true
    livenessEndpoint: "/health?checkType=liveness"
    readinessEnabled: true
    readinessEndpoint: "/health?checkType=readiness"
```
and secondly b/c health endpoint ignores unknown query parameter. it will work with old or new versions of pinot docker images

walterddr commented:
@alvinzhangdd any follow up on the comments above? thanks

alvinzhangdd commented:
@walterddr I will get to it soon. 

Basically I need to add those as optional params into the README, and provide examples like "/health?checkType=liveness"

alvinzhangdd commented:
@walterddr  I addressed the comments, but I had a `Merge branch 'apache:master' into ...` commit, please let me know if I didn't do it correctly and have to close this PR and start a new branch over again.
